### PR TITLE
Refactor ship placement

### PR DESCRIPTION
**Pull request by totine**
- Fleet is generic now
- Ship is an interface now that is implemented by UndirectedShip, DirectedShip and PlacedShip
- Move ship placing from Board to dedicated class (ShipPlacer)
- Board API contains only isMastPlaceable() and placeMast() now

### CODE REVIEW

File: src/main/java/placement/controller/RandomFleetPlacer.java [hunk 1]
```diff
@@ -13,38 +16,41 @@
...
         this.random = new Random();
     }
 
 
-
-    Fleet placeFleet() {
+    Fleet<PlacedShip> placeFleet() {
+        Fleet<PlacedShip> fleetToSend = new Fleet<>();
         for (Ship ship : fleet.getShipList()) {
             while (!ship.isPlaced()) {
```
Comment (boczkas) [inline src/main/java/placement/controller/RandomFleetPlacer.java]: extract method

File: src/main/java/placement/controller/RandomFleetPlacer.java [hunk 2]
```diff
@@ -13,38 +16,41 @@
...
                 Direction direction = getRandomDirection();
-                ship.setDirection(direction);
+                DirectedShip ship2 = ship.direct(direction);
```
Comment (boczkas) [inline src/main/java/placement/controller/RandomFleetPlacer.java]: directedShip

File: src/main/java/placement/controller/RandomFleetPlacer.java [hunk 3]
```diff
@@ -13,38 +16,41 @@
...
                 Coordinates randomCoordinates = getRandomCoordinates(board.rows(), board.cols());
-                board.placeShip(ship, randomCoordinates);
+                ship = ShipPlacer.placeShip(board, ship2, randomCoordinates);
```
Comment (boczkas) [inline src/main/java/placement/controller/RandomFleetPlacer.java]: if(ShipPlacer.placeShipOnBoard(board, directedShip, randomCoordinates) == ShipPlacement.SUCCESSFUL){
  ship = new PlacedShip(ship, randomCoordinates);
}

File: src/main/java/placement/controller/RandomFleetPlacer.java [hunk 4]
```diff
@@ -13,38 +16,41 @@
...
         }
-        return fleet;
+        return fleetToSend;
     }
 
-    public Direction getRandomDirection() {
+    private Direction getRandomDirection() {
+
         Direction[] directions = Direction.class.getEnumConstants();
         int index = random.nextInt(directions.length - 1);
```
Comment (boczkas) [inline src/main/java/placement/controller/RandomFleetPlacer.java]: magic number

File: src/main/java/placement/controller/ShipPlacer.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.List;
+
+/**
+ * utility class with one method for placing one ship on the board
+ */
+public final class ShipPlacer {
+    private ShipPlacer() {
+    }
+
+    static Ship placeShip(Board board, DirectedShip ship, Coordinates headCoordinates) {
```
Comment (boczkas) [inline src/main/java/placement/controller/ShipPlacer.java]: will be changed - no review

File: src/main/java/placement/controller/ShipPlacer.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+        if (mastCoordinates
+                .stream()
+                .allMatch(board::isMastPlaceable)) {
+            mastCoordinates.forEach(board::placeMast);
+            return new PlacedShip(mastCoordinates);
+        }
+        return new UndirectedShip(ship.getMastNumber());
+    }
+
+    public static Ship placeShip(Board board, DirectedShip ship1, int x, int y) {
```
Comment (boczkas) [inline src/main/java/placement/controller/ShipPlacer.java]: ship1 - where is ship2?

Comment (boczkas) [inline src/main/java/placement/controller/ShipPlacer.java]: self comment:
coordinates created only for testing purposes - should be refactored

File: src/main/java/placement/model/Fleet.java [hunk 7]
```diff
@@ -12,28 +14,34 @@
...
-            Ship ship = new Ship(shipType.getSize());
-            shipList.add(ship);
-        }
     }
 
-    public static Fleet createDefaultFleet() {
-        return new Fleet(BATTLESHIP,
+    public static Fleet<Ship> createDefaultFleet() {
+        Fleet<Ship> fleet = new Fleet<>();
+        ShipType[] shipTypes = {BATTLESHIP,
```
Comment (boczkas) [inline src/main/java/placement/model/Fleet.java]: Who are those?

File: src/main/java/placement/model/board/Board.java [hunk 8]
```diff
@@ -45,7 +40,7 @@ public static Board createDefaultBoard() {
      *
      * @return true if a board is not yet populated.
      */
-    boolean isEmpty() {
+    public boolean isEmpty() {
```
Comment (boczkas) [inline src/main/java/placement/model/board/Board.java:44]: self comment:
used only in tests

File: src/main/java/placement/model/board/Board.java [hunk 9]
```diff
@@ -73,56 +68,19 @@ private FieldState getFieldState(Coordinates coordinates) {
         return field.getState();
     }
 
-    FieldState getFieldState(int x, int y) {
+    public FieldState getFieldState(int x, int y) {
```
Comment (boczkas) [inline src/main/java/placement/model/board/Board.java]: self comment:
used only in tests

File: src/main/java/placement/model/board/Board.java [hunk 10]
```diff
@@ -142,6 +100,7 @@ private void surroundWithBuffer(Coordinates mastCoordinates) {
         }
         neighbours.removeIf(this::isFieldOccupied);
         neighbours.removeIf(this::isOutOfBound);
+        neighbours.remove(coordinates);
```
Comment (boczkas) [inline src/main/java/placement/model/board/Board.java]: self comment:
coordinates - coordinatesToSurrounded with buffer

File: src/main/java/placement/model/ship/DirectedShip.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+/**
+ * Represents a ship, that is directed but not placed in the context of placing ships on a board.
+ * Knows its direction and mast number.
+ */
+public class DirectedShip implements Ship {
+    private final int mastNumber;
+    private Direction direction;
+
+
```
Comment (boczkas) [inline src/main/java/placement/model/ship/DirectedShip.java]: /n to be removed

File: src/main/java/placement/model/ship/DirectedShip.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+    public DirectedShip(int mastCount, Direction direction) {
+        this.mastNumber = mastCount;
```
Comment (boczkas) [inline src/main/java/placement/model/ship/DirectedShip.java]: both of them are the same, you can have same name

File: src/main/java/placement/model/ship/DirectedShip.java [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        return this;
+    }
+
+
+    public final List<Coordinates> getTempCoordinates(Coordinates headCoordinates) {
+        List<Coordinates> mastCoordinates = new ArrayList<>();
+        mastCoordinates.add(headCoordinates);
+        Coordinates currentCoordinates = headCoordinates;
+        while (mastCoordinates.size() < mastNumber) {
+            Coordinates nextCoord = direction.nextCoordinates();
```
Comment (boczkas) [inline src/main/java/placement/model/ship/DirectedShip.java]: Coordinates nextCoord initialization before while

File: src/main/java/placement/model/ship/PlacedShip.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+import placement.model.Coordinates;
+
+import java.util.List;
+import java.util.UUID;
+
+/**
+ * Represents a placed ship in the context of placing ships on a board. Knows its absolute mast coordinates and id.
+ */
+public final class PlacedShip implements Ship {
+    private final String id;
```
Comment (boczkas) [inline src/main/java/placement/model/ship/PlacedShip.java]: remove it

File: src/main/java/placement/model/ship/UndirectedShip.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    public UndirectedShip(int mastNumber) {
+        this.mastNumber = mastNumber;
+    }
+
+    @Override
+    public final boolean isPlaced() {
+        return false;
+    }
+
+
```
Comment (boczkas) [inline src/main/java/placement/model/ship/UndirectedShip.java]: /n remove

File: src/test/java/placement/model/ShipPlacerTests.java [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+    private void createBoard() {
+        board = Board.createDefaultBoard();
+    }
+
+    @AfterMethod
+    private void printBoard() {
+        System.out.println(board);
+    }
+
+
```
Comment (boczkas) [inline src/test/java/placement/model/ShipPlacerTests.java:27]: /n remove

File: src/test/java/placement/model/ShipPlacerTests.java [hunk 17]
```diff
@@ -0,0 +1,287 @@
+package placement.model;
```
Comment (boczkas) [inline src/test/java/placement/model/ShipPlacerTests.java:1]: try to extract magic numbers

File: src/test/java/placement/model/ShipPlacerTests.java [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+                {7, 4}, {7, 5}, {7, 6}, {7, 7}, {7, 8}};
+        multipleAssertEqualsIsExpectedFieldState(board, FieldState.BUFFER, coordsThatShouldBeBufferForShip2);
+
+
+    }
+
+    @Test
+    public void givenEmptyBoard_WhenPLacingTwoShipsNextToEachOther_ThenEverythingOK() {
+
+        DirectedShip ship1 = new DirectedShip(3, Direction.EAST);
```
Comment (boczkas) [inline src/test/java/placement/model/ShipPlacerTests.java]: ship1 - shipDirectedEast
ship2 - shipDirectedSouth

**Review by boczkas — CHANGES_REQUESTED** [review]
Cosmetics

File: src/main/java/placement/model/board/Board.java [hunk 19]
```diff
@@ -88,19 +85,19 @@ private void surroundWithBuffer(Coordinates mastCoordinates) {
         neighbours.forEach(boardFields::markAsBuffer);
     }
 
-    private Set<Coordinates> getNeighboursToMarkAsBuffer(Coordinates coordinates) {
+    private Set<Coordinates> getNeighboursToMarkAsBuffer(Coordinates coordinatesToSurrounded) {
```
Comment (boczkas) [inline src/main/java/placement/model/board/Board.java]: self comment:
name to be changed

**Review by boczkas — APPROVED** [review]
:+1:

File: src/main/java/placement/controller/RandomFleetPlacer.java [hunk 20]
```diff
@@ -13,38 +17,54 @@
...
+    RandomFleetPlacer(Fleet<UndirectedShip> fleet, Board board) {
         this.fleet = fleet;
         this.board = board;
         this.random = new Random();
     }
 
 
-
-    Fleet placeFleet() {
+    Fleet<PlacedShip> placeFleet() {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/controller/RandomFleetPlacer.java:32]: What about getPlacedFleet?

File: src/main/java/placement/controller/ShipPlacer.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+import java.util.List;
+
+/**
+ * utility class with one method for placing one ship on the board
+ */
+public final class ShipPlacer {
+    private ShipPlacer() {
+    }
+
+    public static boolean tryToPlaceShip(Board board, DirectedShip ship, Coordinates headCoordinates) {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/controller/ShipPlacer.java:16]: javadoc?

File: src/main/java/placement/controller/FleetController.java [hunk 22]
```diff
@@ -2,14 +2,18 @@
...
 
 /**
  * contains methods related to fleet control.
  */
 public final class FleetController {
     private FleetController() { }
-    public static Fleet generatePlacedStandardFleet() {
-        Fleet fleet = Fleet.createDefaultFleet();
+
+    public static Fleet<PlacedShip> generatePlacedStandardFleet() {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/controller/FleetController.java:15]: javadoc?

File: src/main/java/placement/model/Fleet.java [hunk 23]
```diff
@@ -2,38 +2,46 @@
...
-
+public final class Fleet<T extends Ship> {
+    private List<T> shipList;
 
-    private Fleet(ShipType... shipTypes) {
+    public Fleet() {
         shipList = new ArrayList<>();
+    }
+
+    public static Fleet<UndirectedShip> createDefaultFleet() {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/model/Fleet.java:27]: javadoc?

File: src/main/java/placement/model/board/Board.java [hunk 24]
```diff
@@ -68,80 +64,40 @@ public int cols() {
...
         HashSet<Coordinates> neighbours = new HashSet<>();
         for (int i = -1; i <= 1; i++) {
             for (int j = -1; j <= 1; j++) {
                 Coordinates neighbour = new Coordinates(
-                        coordinates.getX() + i,
-                        coordinates.getY() + j);
+                        coordinatesToSurrounded.getX() + i,
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/model/board/Board.java]: You have already a method to add coordinates to each other (in Coordinates class)

File: src/main/java/placement/model/board/Board.java [hunk 25]
```diff
@@ -68,80 +64,40 @@ public int cols() {
...
 
     private void surroundWithBuffer(Coordinates mastCoordinates) {
         Set<Coordinates> neighbours = getNeighboursToMarkAsBuffer(mastCoordinates);
         neighbours.forEach(boardFields::markAsBuffer);
     }
 
-    private Set<Coordinates> getNeighboursToMarkAsBuffer(Coordinates coordinates) {
+    private Set<Coordinates> getNeighboursToMarkAsBuffer(Coordinates coordinatesToSurrounded) {
         HashSet<Coordinates> neighbours = new HashSet<>();
         for (int i = -1; i <= 1; i++) {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/model/board/Board.java]: What about extracting it to a new method?

Comment (KrzysztofDzioba) [inline src/main/java/placement/model/board/Board.java]: You can change "i" for "row"

File: src/main/java/placement/model/board/Board.java [hunk 26]
```diff
@@ -68,80 +64,40 @@ public int cols() {
...
             for (int j = -1; j <= 1; j++) {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/model/board/Board.java]: and "j" for "col"

File: src/main/java/placement/view/FleetFromRandomGenerator.java [hunk 27]
```diff
@@ -2,18 +2,19 @@
...
-import placement.model.ship.Ship;
+import placement.model.ship.PlacedShip;
+
 
 /**
  * GUI-side fleet randomizer
  */
 public class FleetFromRandomGenerator implements FleetDAO {
     @Override
     public final GUIFleet getFleet() {
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/view/FleetFromRandomGenerator.java]: what about getGUIFleet?

File: src/main/java/placement/view/GUIShip.java [hunk 28]
```diff
@@ -8,19 +8,15 @@
  * GUI representation of ship object
  */
 class GUIShip {
-    private String id;
     private List<Coordinates> structure;
```
Comment (KrzysztofDzioba) [inline src/main/java/placement/view/GUIShip.java]: structure -> coordinates

**Review by KrzysztofDzioba — CHANGES_REQUESTED** [review]
Small fixes needed

**Review by KrzysztofDzioba — APPROVED** [review]

**Pull request merged**